README.md
# Ecto.ULID Next

[![Elixir CI](https://github.com/woylie/ecto_ulid/actions/workflows/elixir.yml/badge.svg)](https://github.com/woylie/ecto_ulid/actions/workflows/elixir.yml)
[![Hex](https://img.shields.io/hexpm/v/ecto_ulid_next)](https://hex.pm/packages/ecto_ulid_next)

Ecto.ULID Next is an `Ecto.Type` implementation for [ULID](https://github.com/ulid/spec).

Originally forked from [TheRealReal/ecto-ulid](https://github.com/TheRealReal/ecto-ulid),
this library is actively maintained to ensure compatibility with both current
and upcoming Ecto releases.

## Features

- Generate ULIDs in either Base32 or binary format.
- Generate ULIDs based on a specific timestamp.
- Autogenerate ULIDs when used as a primary key in your schema.
- Read and write ULIDs using the database's native `uuid` type—no need for
  additional database extensions.
- Supports Ecto 3.2 and above.
- Supports the [officially supported Elixir versions](https://hexdocs.pm/elixir/compatibility-and-deprecations.html) (currently ~> 1.11).
- Tested and confirmed to work on both PostgreSQL and MySQL.
- Optimized for high throughput scenarios.

## Why Choose ULID?

ULID (Universally Unique Lexicographically Sortable Identifier) offers
lexicographic sorting, unlike UUID. This feature simplifies chronological
sorting and range queries, making ULID advantageous for time-series data, event
sourcing, and other specific use-cases.

Note: UUID v7 is a new proposed format that also aims to offer lexicographic
sorting.

## Compatibility Considerations

ULID is binary-compatible with UUID, which means `Ecto.ULID` can be used
interchangeably in environments where `Ecto.UUID` is supported. It has been
confirmed to work with PostgreSQL and MySQL with the `uuid` column type.

## Performance

`Ecto.ULID` is optimized for quick operations, particularly useful in scenarios
requiring the handling of a large volume of events. The library borrows
techniques from `Ecto.UUID` to achieve sub-microsecond times for most
operations.

A benchmark suite is included to test the performance. Clone the repository and
run `mix run bench/ulid_bench.exs` to gauge how the library performs on your own
system.

## Installation

To install `ecto_ulid_next`, add it to your project's dependencies in `mix.exs`:

```elixir
defp deps do
  [
    {:ecto_ulid_next, "~> 1.0.2"}
  ]
end
```

## Usage Scenarios

The following sections outline three practical examples of how to use
`Ecto.ULID`. For API details, refer to the
[documentation](https://hexdocs.pm/ecto_ulid_next).

### Using `Ecto.ULID` as a Primary Key or Foreign Key for a Single Table

#### Migration

In your migration file, specify `:binary_id` as the column type for the
primary key and any foreign keys.

```elixir
create table(:events, primary_key: false) do
  add :id, :binary_id, primary_key: true
  add :user_id, references(:users, type: :binary_id)
  # additional fields
end
```

#### Schema Definition

In your schema, set both `@primary_key` and `@foreign_key_type` to `Ecto.ULID`.
This will default both the primary key and all foreign keys in the schema to use
the ULID type.

```elixir
defmodule MyApp.Event do
  use Ecto.Schema

  @primary_key {:id, Ecto.ULID, autogenerate: true}
  @foreign_key_type Ecto.ULID

  schema "events" do
    belongs_to :user, MyApp.User
    # additional fields
  end
end
```

To specify the ULID type for an individual association within the current
schema, you can set the `type` option in the `belongs_to` function:

```elixir
defmodule MyApp.Event do
  use Ecto.Schema

  @primary_key {:id, Ecto.ULID, autogenerate: true}

  schema "events" do
    belongs_to :user, MyApp.User, type: Ecto.ULID
    # additional fields
  end
end
```

### Using `Ecto.ULID` as Primary Keys Across Multiple Tables

#### Global Migration Configuration

First, set a global configuration for all migrations to use `:binary_id` as the
primary key type in `config/config.exs`.

```elixir
config :my_app, MyApp.Repo,
  migration_primary_key: [name: :id, type: :binary_id],
  migration_foreign_key: [type: :binary_id]
```

With this configuration, there's no need to explicitly define the primary key
column and the foreign key column type.

```elixir
create table(:events) do
  belongs_to :user, MyApp.User
  # additional fields
end
```

#### Shared Schema Configuration

Create a shared module to house general schema configurations. This module will
set `Ecto.ULID` as the default for both primary and foreign keys.

```elixir
defmodule MyApp.Schema do
  @moduledoc false

  defmacro __using__(_) do
    quote do
      use Ecto.Schema

      @primary_key {:id, Ecto.ULID, autogenerate: true}
      @foreign_key_type Ecto.ULID
    end
  end
end
```

When defining individual schemas, simply use `MyApp.Schema` instead of
`Ecto.Schema`. This eliminates the need to specify these attributes in each
schema.

```elixir
defmodule MyApp.Event do
  use MyApp.Schema

  schema "events" do
    # additional fields
  end
end
```

### Manual ULID Generation

#### Using the Current System Time

Use `Ecto.ULID.generate/0` or `Ecto.ULID.bingenerate/0` to create ULIDs in
string or binary format based on the current system time. This is useful for
sending ULIDs to external systems or for use with `c:Ecto.Repo.insert_all/3`,
where autogenerated fields are not set automatically.

```elixir
iex> Ecto.ULID.generate()
"01BZ13RV29T5S8HV45EDNC748P"

iex> Ecto.ULID.bingenerate()
<<1, 95, 194, 60, 108, 73, 209, 114, 136, 236, 133, 115, 106, 195, 145, 22>>
```

#### Specifying a Timestamp

To generate a ULID based on a specific timestamp—useful for backfilling
historical data—you can use `Ecto.ULID.generate/1` or `Ecto.ULID.bingenerate/1`.

```elixir
iex> Ecto.ULID.generate(1402899630000)
"018THNB1XGQZ7T929PD126SM3Z"

iex> Ecto.ULID.bingenerate(1402899630000)
<<1, 70, 163, 85, 135, 176, 12, 219, 112, 32, 157, 209, 98, 152, 55, 37>>
```

## Further Reading

### Specifications

- [ULID specification](https://github.com/ulid/spec)
- [UUID v7 specification](https://datatracker.ietf.org/doc/html/draft-ietf-uuidrev-rfc4122bis#name-uuid-version-7)

### Database Support and Tools

ULIDs are not natively supported by most databases, which means querying data
directly may display IDs in UUID format rather than the Crockford Base32
representation. For PostgreSQL, the following tool can help convert between ULID
and UUID formats:

- [pgsql-ulid](https://github.com/scoville/pgsql-ulid)

### Articles and Insights

- [Understanding UUIDs, ULIDs and String Representations](https://sudhir.io/uuids-ulids)


CHANGELOG.md
# Changelog

## Unreleased

## [1.0.2] - 2023-09-09

### Changed

- Documentation improvements.

## [1.0.1] - 2022-12-14

### Fixed

- Source references in documentation.

## [1.0.0] - 2022-12-14

Stable release.

### Changed

- Minor dependency and documentation updates.

## [0.3.1] - 2022-02-22

Maintenance release

### Changed

- Elixir version support now follows the
  [officially supported Elixir versions](https://hexdocs.pm/elixir/1.13/compatibility-and-deprecations.html).
- Update test matrix accordingly.
- Improve documentation and type specs.

## [0.3.0] - 2020-11-16

### Changed

- Minimum supported Ecto is now 3.2.
- Minimum supported Elixir is now 1.7.

## [0.2.0] - 2019-01-17

### Changed

- Minimum supported Elixir is now 1.4.
- ([#3](https://github.com/TheRealReal/ecto-ulid/pull/3))
  Fix deprecation warnings regarding time units.

## [0.1.1] - 2018-12-03

### Added

- ([#2](https://github.com/TheRealReal/ecto-ulid/pull/2))
  Add support for Ecto 3.x.

## [0.1.0] - 2018-06-06

### Added

- Generate ULID in Base32 or binary format.
- Generate ULID for a given timestamp.
- Autogenerate ULID when used as a primary key.
- Supports reading and writing ULID in a database backed by its native `uuid` type (no database
  extensions required).
- Supports Ecto 2.x.
- Supports Elixir 1.2 and newer.
- Tested with PostgreSQL and MySQL.


test/ecto/ulid_test.exs
defmodule Ecto.ULIDTest do
  use ExUnit.Case, async: true

  @binary <<1, 95, 194, 60, 108, 73, 209, 114, 136, 236, 133, 115, 106, 195,
            145, 22>>
  @encoded "01BZ13RV29T5S8HV45EDNC748P"

  describe "type/0" do
    test "returns :uuid" do
      assert Ecto.ULID.type() == :uuid
    end
  end

  describe "generate/0" do
    test "generate/0 encodes milliseconds in first 10 characters" do
      # test case from ULID README: https://github.com/ulid/javascript#seed-time
      <<encoded::bytes-size(10), _rest::bytes-size(16)>> =
        Ecto.ULID.generate(1_469_918_176_385)

      assert encoded == "01ARYZ6S41"
    end

    test "generate/0 generates unique identifiers" do
      ulid1 = Ecto.ULID.generate()
      ulid2 = Ecto.ULID.generate()

      assert ulid1 != ulid2
    end
  end

  describe "bingenerate/0" do
    test "encodes milliseconds in first 48 bits" do
      now = System.system_time(:millisecond)
      <<time::48, _random::80>> = Ecto.ULID.bingenerate()

      assert_in_delta now, time, 10
    end

    test "generates unique identifiers" do
      ulid1 = Ecto.ULID.bingenerate()
      ulid2 = Ecto.ULID.bingenerate()

      assert ulid1 != ulid2
    end
  end

  # cast/1

  describe "cast/1" do
    test "returns valid ULID" do
      {:ok, ulid} = Ecto.ULID.cast(@encoded)
      assert ulid == @encoded
    end

    test "returns ULID for encoding of correct length" do
      {:ok, ulid} = Ecto.ULID.cast("00000000000000000000000000")
      assert ulid == "00000000000000000000000000"
    end

    test "returns error when encoding is too short" do
      assert Ecto.ULID.cast("0000000000000000000000000") == :error
    end

    test "returns error when encoding is too long" do
      assert Ecto.ULID.cast("000000000000000000000000000") == :error
    end

    test "returns error when encoding contains letter I" do
      assert Ecto.ULID.cast("I0000000000000000000000000") == :error
    end

    test "returns error when encoding contains letter L" do
      assert Ecto.ULID.cast("L0000000000000000000000000") == :error
    end

    test "returns error when encoding contains letter O" do
      assert Ecto.ULID.cast("O0000000000000000000000000") == :error
    end

    test "returns error when encoding contains letter U" do
      assert Ecto.ULID.cast("U0000000000000000000000000") == :error
    end

    test "returns error for invalid encoding" do
      assert Ecto.ULID.cast("$0000000000000000000000000") == :error
    end
  end

  describe "cast!/1" do
    test "returns valid ULID" do
      assert Ecto.ULID.cast!(@encoded) == @encoded
    end

    test "returns error when encoding is too short" do
      assert_raise Ecto.CastError, fn ->
        Ecto.ULID.cast!("0000000000000000000000000")
      end
    end
  end

  describe "dump/1" do
    test "dumps valid ULID to binary" do
      {:ok, bytes} = Ecto.ULID.dump(@encoded)
      assert bytes == @binary
    end

    test "dumps encoding of correct length" do
      {:ok, bytes} = Ecto.ULID.dump("00000000000000000000000000")
      assert bytes == <<0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0>>
    end

    test "returns error when encoding is too short" do
      assert Ecto.ULID.dump("0000000000000000000000000") == :error
    end

    test "returns error when encoding is too long" do
      assert Ecto.ULID.dump("000000000000000000000000000") == :error
    end

    test "returns error when encoding contains letter I" do
      assert Ecto.ULID.dump("I0000000000000000000000000") == :error
    end

    test "returns error when encoding contains letter L" do
      assert Ecto.ULID.dump("L0000000000000000000000000") == :error
    end

    test "returns error when encoding contains letter O" do
      assert Ecto.ULID.dump("O0000000000000000000000000") == :error
    end

    test "returns error when encoding contains letter U" do
      assert Ecto.ULID.dump("U0000000000000000000000000") == :error
    end

    test "returns error for invalid encoding" do
      assert Ecto.ULID.dump("$0000000000000000000000000") == :error
    end
  end

  describe "load/1" do
    test "encodes binary as ULID" do
      {:ok, encoded} = Ecto.ULID.load(@binary)
      assert encoded == @encoded
    end

    test "encodes binary of correct length" do
      {:ok, encoded} =
        Ecto.ULID.load(<<0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0>>)

      assert encoded == "00000000000000000000000000"
    end

    test "returns error when data is too short" do
      assert Ecto.ULID.load(<<0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0>>) ==
               :error
    end

    test "returns error when data is too long" do
      assert Ecto.ULID.load(
               <<0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0>>
             ) ==
               :error
    end
  end

  describe "autogenerate/0" do
    test "returns a valid ULID" do
      ulid = Ecto.ULID.autogenerate()
      assert {:ok, ^ulid} = Ecto.ULID.cast(ulid)
    end
  end
end


mix.exs
defmodule Ecto.ULID.Mixfile do
  use Mix.Project

  @source_url "https://github.com/woylie/ecto_ulid"
  @version "1.0.2"

  def project do
    [
      app: :ecto_ulid_next,
      version: @version,
      elixir: "~> 1.10",
      start_permanent: Mix.env() == :prod,
      deps: deps(),
      test_coverage: [tool: ExCoveralls],
      preferred_cli_env: [
        coveralls: :test,
        "coveralls.detail": :test,
        "coveralls.post": :test,
        "coveralls.html": :test,
        "coveralls.github": :test
      ],
      name: "Ecto.ULID",
      description: "An Ecto.Type implementation of ULID.",
      package: package(),
      source_url: @source_url,
      homepage_url: @source_url,
      docs: docs(),
      dialyzer: [
        plt_file: {:no_warn, ".plts/dialyzer.plt"}
      ]
    ]
  end

  def application do
    [
      extra_applications: [:logger]
    ]
  end

  defp deps do
    [
      {:benchee, "~> 1.0", only: :dev},
      {:credo, "~> 1.5", only: [:dev, :test], runtime: false},
      {:dialyxir, "~> 1.4.1", only: [:dev], runtime: false},
      {:ecto, "~> 3.2"},
      {:ex_doc, "~> 0.16", only: :dev, runtime: false},
      {:excoveralls, "~> 0.10", only: :test}
    ]
  end

  defp package do
    [
      maintainers: ["Mathias Polligkeit"],
      licenses: ["MIT"],
      links: %{
        "GitHub" => @source_url,
        "Changelog" => @source_url <> "/blob/main/CHANGELOG.md"
      },
      files: ~w(lib .formatter.exs mix.exs README* LICENSE* CHANGELOG*)
    ]
  end

  defp docs do
    [
      main: "readme",
      extras: ["README.md", "CHANGELOG.md"],
      source_ref: "v" <> @version,
      skip_undefined_reference_warnings_on: ["CHANGELOG.md"]
    ]
  end
end


.credo.exs
# This file contains the configuration for Credo and you are probably reading
# this after creating it with `mix credo.gen.config`.
#
# If you find anything wrong or unclear in this file, please report an
# issue on GitHub: https://github.com/rrrene/credo/issues
#
%{
  #
  # You can have as many configs as you like in the `configs:` field.
  configs: [
    %{
      #
      # Run any config using `mix credo -C <name>`. If no config name is given
      # "default" is used.
      #
      name: "default",
      #
      # These are the files included in the analysis:
      files: %{
        #
        # You can give explicit globs or simply directories.
        # In the latter case `**/*.{ex,exs}` will be used.
        #
        included: [
          "bench",
          "lib/",
          "test/"
        ],
        excluded: [~r"/_build/", ~r"/deps/", ~r"/node_modules/"]
      },
      #
      # Load and configure plugins here:
      #
      plugins: [],
      #
      # If you create your own checks, you must specify the source files for
      # them here, so they can be loaded by Credo before running the analysis.
      #
      requires: [],
      #
      # If you want to enforce a style guide and need a more traditional linting
      # experience, you can change `strict` to `true` below:
      #
      strict: true,
      #
      # To modify the timeout for parsing files, change this value:
      #
      parse_timeout: 5000,
      #
      # If you want to use uncolored output by default, you can change `color`
      # to `false` below:
      #
      color: true,
      #
      # You can customize the parameters of any check by adding a second element
      # to the tuple.
      #
      # To disable a check put `false` as second element:
      #
      #     {Credo.Check.Design.DuplicatedCode, false}
      #
      checks: %{
        enabled: [
          #
          ## Consistency Checks
          #
          {Credo.Check.Consistency.ExceptionNames, []},
          {Credo.Check.Consistency.LineEndings, []},
          {Credo.Check.Consistency.ParameterPatternMatching, []},
          {Credo.Check.Consistency.SpaceAroundOperators, []},
          {Credo.Check.Consistency.SpaceInParentheses, []},
          {Credo.Check.Consistency.TabsOrSpaces, []},

          #
          ## Design Checks
          #
          # You can customize the priority of any check
          # Priority values are: `low, normal, high, higher`
          #
          {Credo.Check.Design.AliasUsage,
           [
             priority: :low,
             if_nested_deeper_than: 2,
             if_called_more_often_than: 0
           ]},
          # You can also customize the exit_status of each check.
          # If you don't want TODO comments to cause `mix credo` to fail, just
          # set this value to 0 (zero).
          #
          {Credo.Check.Design.TagTODO, [exit_status: 2]},
          {Credo.Check.Design.TagFIXME, []},

          #
          ## Readability Checks
          #
          {Credo.Check.Readability.AliasOrder, []},
          {Credo.Check.Readability.FunctionNames, []},
          {Credo.Check.Readability.LargeNumbers, []},
          {Credo.Check.Readability.MaxLineLength, [priority: :low, max_length: 80]},
          {Credo.Check.Readability.ModuleAttributeNames, []},
          {Credo.Check.Readability.ModuleDoc, []},
          {Credo.Check.Readability.ModuleNames, []},
          {Credo.Check.Readability.ParenthesesInCondition, []},
          {Credo.Check.Readability.ParenthesesOnZeroArityDefs, []},
          {Credo.Check.Readability.PipeIntoAnonymousFunctions, []},
          {Credo.Check.Readability.PredicateFunctionNames, []},
          {Credo.Check.Readability.PreferImplicitTry, []},
          {Credo.Check.Readability.RedundantBlankLines, []},
          {Credo.Check.Readability.Semicolons, []},
          {Credo.Check.Readability.SpaceAfterCommas, []},
          {Credo.Check.Readability.StringSigils, []},
          {Credo.Check.Readability.TrailingBlankLine, []},
          {Credo.Check.Readability.TrailingWhiteSpace, []},
          {Credo.Check.Readability.UnnecessaryAliasExpansion, []},
          {Credo.Check.Readability.VariableNames, []},
          {Credo.Check.Readability.WithSingleClause, []},

          #
          ## Refactoring Opportunities
          #
          {Credo.Check.Refactor.Apply, []},
          {Credo.Check.Refactor.CondStatements, []},
          {Credo.Check.Refactor.FunctionArity, []},
          {Credo.Check.Refactor.LongQuoteBlocks, []},
          {Credo.Check.Refactor.MatchInCondition, []},
          {Credo.Check.Refactor.MapJoin, []},
          {Credo.Check.Refactor.NegatedConditionsInUnless, []},
          {Credo.Check.Refactor.NegatedConditionsWithElse, []},
          {Credo.Check.Refactor.Nesting, []},
          {Credo.Check.Refactor.UnlessWithElse, []},
          {Credo.Check.Refactor.WithClauses, []},
          {Credo.Check.Refactor.FilterCount, []},
          {Credo.Check.Refactor.FilterFilter, []},
          {Credo.Check.Refactor.RejectReject, []},
          {Credo.Check.Refactor.RedundantWithClauseResult, []},

          #
          ## Warnings
          #
          {Credo.Check.Warning.ApplicationConfigInModuleAttribute, []},
          {Credo.Check.Warning.BoolOperationOnSameValues, []},
          {Credo.Check.Warning.Dbg, []},
          {Credo.Check.Warning.ExpensiveEmptyEnumCheck, []},
          {Credo.Check.Warning.IExPry, []},
          {Credo.Check.Warning.IoInspect, []},
          {Credo.Check.Warning.MissedMetadataKeyInLoggerConfig, []},
          {Credo.Check.Warning.OperationOnSameValues, []},
          {Credo.Check.Warning.OperationWithConstantResult, []},
          {Credo.Check.Warning.RaiseInsideRescue, []},
          {Credo.Check.Warning.SpecWithStruct, []},
          {Credo.Check.Warning.WrongTestFileExtension, []},
          {Credo.Check.Warning.UnusedEnumOperation, []},
          {Credo.Check.Warning.UnusedFileOperation, []},
          {Credo.Check.Warning.UnusedKeywordOperation, []},
          {Credo.Check.Warning.UnusedListOperation, []},
          {Credo.Check.Warning.UnusedPathOperation, []},
          {Credo.Check.Warning.UnusedRegexOperation, []},
          {Credo.Check.Warning.UnusedStringOperation, []},
          {Credo.Check.Warning.UnusedTupleOperation, []},
          {Credo.Check.Warning.UnsafeExec, []},

          #
          ## Checks disabled by default
          #


          {Credo.Check.Readability.SinglePipe, []},
          {Credo.Check.Readability.StrictModuleLayout, []},
          {Credo.Check.Readability.WithCustomTaggedTuple, []},
          {Credo.Check.Refactor.DoubleBooleanNegation, []},
          {Credo.Check.Refactor.FilterReject, []},
          {Credo.Check.Refactor.MapMap, []},
          {Credo.Check.Refactor.PipeChainStart, []},
          {Credo.Check.Refactor.RejectFilter, []},
          {Credo.Check.Warning.MapGetUnsafePass, []}
        ],
        disabled: [
          #
          # Checks scheduled for next check update (opt-in for now, just replace `false` with `[]`)

          #
          # Controversial and experimental checks (opt-in, just move the check to `:enabled`
          #   and be sure to use `mix credo --strict` to see low priority checks)
          #
          {Credo.Check.Consistency.MultiAliasImportRequireUse, []},
          {Credo.Check.Consistency.UnusedVariableNames, []},
          {Credo.Check.Design.DuplicatedCode, []},
          {Credo.Check.Design.SkipTestWithoutComment, []},
          {Credo.Check.Readability.AliasAs, []},
          {Credo.Check.Readability.BlockPipe, []},
          {Credo.Check.Readability.ImplTrue, []},
          {Credo.Check.Readability.MultiAlias, []},
          {Credo.Check.Readability.NestedFunctionCalls, []},
          {Credo.Check.Readability.OneArityFunctionInPipe, []},
          {Credo.Check.Readability.OnePipePerLine, []},
          {Credo.Check.Readability.SeparateAliasRequire, []},
          {Credo.Check.Readability.SingleFunctionToBlockPipe, []},
          {Credo.Check.Readability.Specs, []},
          {Credo.Check.Refactor.ABCSize, []},
          {Credo.Check.Refactor.AppendSingleItem, []},
          {Credo.Check.Refactor.CyclomaticComplexity, []},
          {Credo.Check.Refactor.IoPuts, []},
          {Credo.Check.Refactor.ModuleDependencies, []},
          {Credo.Check.Refactor.NegatedIsNil, []},
          {Credo.Check.Refactor.PassAsyncInTestCases, []},
          {Credo.Check.Refactor.VariableRebinding, []},
          {Credo.Check.Warning.LazyLogging, []},
          {Credo.Check.Warning.LeakyEnvironment, []},
          {Credo.Check.Warning.MixEnv, []},
          {Credo.Check.Warning.UnsafeToAtom, []}

          # {Credo.Check.Refactor.MapInto, []},

          #
          # Custom checks can be created using `mix credo.gen.check`.
          #
        ]
      }
    }
  ]
}


lib/ecto/ulid.ex
defmodule Ecto.ULID do
  @moduledoc """
  An Ecto type for ULID strings.
  """

  use Ecto.Type

  @typedoc """
  A hex-encoded ULID string.
  """
  @type t :: <<_::208>>

  @doc """
  The underlying schema type (`:uuid`).
  """
  @impl Ecto.Type
  def type, do: :uuid

  @doc """
  Casts a string to ULID.

  ## Example

      iex> Ecto.ULID.cast("018THNB1XG1KDQ084XT5H9GDS5")
      {:ok, "018THNB1XG1KDQ084XT5H9GDS5"}

      iex> Ecto.ULID.cast("abc")
      :error
  """
  @impl Ecto.Type
  def cast(<<_::bytes-size(26)>> = value) do
    if valid?(value) do
      {:ok, value}
    else
      :error
    end
  end

  def cast(_), do: :error

  @doc """
  Same as `cast/1`, but raises `Ecto.CastError` on invalid arguments.

  ## Example

      iex> Ecto.ULID.cast!("018THNB1XG1KDQ084XT5H9GDS5")
      "018THNB1XG1KDQ084XT5H9GDS5"

      iex> Ecto.ULID.cast!("abc")
      ** (Ecto.CastError) cannot cast "abc" to Ecto.ULID
  """
  def cast!(value) do
    case cast(value) do
      {:ok, ulid} -> ulid
      :error -> raise Ecto.CastError, type: __MODULE__, value: value
    end
  end

  @doc """
  Converts a Crockford Base32 encoded ULID into a binary.

  ## Example

      iex> Ecto.ULID.dump("018THNB1XG1KDQ084XT5H9GDS5")
      {:ok, <<1, 70, 163, 85, 135, 176, 12, 219, 112, 32, 157, 209, 98, 152, 55, 37>>}
  """
  @impl Ecto.Type
  def dump(<<_::bytes-size(26)>> = encoded), do: decode(encoded)
  def dump(_), do: :error

  @doc """
  Converts a binary ULID into a Crockford Base32 encoded string.

  ## Example

      iex> Ecto.ULID.load(<<1, 70, 163, 85, 135, 176, 12, 219, 112, 32, 157, 209, 98, 152, 55, 37>>)
      {:ok, "018THNB1XG1KDQ084XT5H9GDS5"}
  """
  @impl Ecto.Type
  def load(<<_::unsigned-size(128)>> = bytes), do: encode(bytes)
  def load(_), do: :error

  @impl Ecto.Type
  def autogenerate, do: generate()

  @doc """
  Generates a Crockford Base32 encoded ULID.

  Optionally takes a Unix timestamp with millisecond precision as an argument.
  If no argument is passed, the current time is used as a base for the ULID.

  ## Examples

      iex> Ecto.ULID.generate()
      "01FWEPCVGGMQQ0YTKS7BYFQH46"

      iex> Ecto.ULID.generate(1402899630000)
      "018THNB1XGQZ7T929PD126SM3Z"
  """
  @spec generate(integer) :: binary
  def generate(timestamp \\ System.system_time(:millisecond)) do
    {:ok, ulid} = encode(bingenerate(timestamp))
    ulid
  end

  @doc """
  Generates a binary ULID.

  Optionally takes a Unix timestamp with millisecond precision as an argument.
  If no argument is passed, the current time is used as a base for the ULID.

  ## Examples

      iex> Ecto.ULID.bingenerate()
      <<1, 127, 29, 107, 117, 125, 183, 154, 147, 224, 191, 135, 161, 183, 2, 52>>

      iex> Ecto.ULID.bingenerate(1402899630000)
      <<1, 70, 163, 85, 135, 176, 12, 219, 112, 32, 157, 209, 98, 152, 55, 37>>
  """
  @spec bingenerate(integer) :: binary
  def bingenerate(timestamp \\ System.system_time(:millisecond)) do
    <<timestamp::unsigned-size(48), :crypto.strong_rand_bytes(10)::binary>>
  end

  defp encode(
         <<b1::3, b2::5, b3::5, b4::5, b5::5, b6::5, b7::5, b8::5, b9::5,
           b10::5, b11::5, b12::5, b13::5, b14::5, b15::5, b16::5, b17::5,
           b18::5, b19::5, b20::5, b21::5, b22::5, b23::5, b24::5, b25::5,
           b26::5>>
       ) do
    <<e(b1), e(b2), e(b3), e(b4), e(b5), e(b6), e(b7), e(b8), e(b9), e(b10),
      e(b11), e(b12), e(b13), e(b14), e(b15), e(b16), e(b17), e(b18), e(b19),
      e(b20), e(b21), e(b22), e(b23), e(b24), e(b25), e(b26)>>
  catch
    :error -> :error
  else
    encoded -> {:ok, encoded}
  end

  defp encode(_), do: :error

  @compile {:inline, e: 1}

  defp e(0), do: ?0
  defp e(1), do: ?1
  defp e(2), do: ?2
  defp e(3), do: ?3
  defp e(4), do: ?4
  defp e(5), do: ?5
  defp e(6), do: ?6
  defp e(7), do: ?7
  defp e(8), do: ?8
  defp e(9), do: ?9
  defp e(10), do: ?A
  defp e(11), do: ?B
  defp e(12), do: ?C
  defp e(13), do: ?D
  defp e(14), do: ?E
  defp e(15), do: ?F
  defp e(16), do: ?G
  defp e(17), do: ?H
  defp e(18), do: ?J
  defp e(19), do: ?K
  defp e(20), do: ?M
  defp e(21), do: ?N
  defp e(22), do: ?P
  defp e(23), do: ?Q
  defp e(24), do: ?R
  defp e(25), do: ?S
  defp e(26), do: ?T
  defp e(27), do: ?V
  defp e(28), do: ?W
  defp e(29), do: ?X
  defp e(30), do: ?Y
  defp e(31), do: ?Z

  defp decode(
         <<c1::8, c2::8, c3::8, c4::8, c5::8, c6::8, c7::8, c8::8, c9::8,
           c10::8, c11::8, c12::8, c13::8, c14::8, c15::8, c16::8, c17::8,
           c18::8, c19::8, c20::8, c21::8, c22::8, c23::8, c24::8, c25::8,
           c26::8>>
       ) do
    <<d(c1)::3, d(c2)::5, d(c3)::5, d(c4)::5, d(c5)::5, d(c6)::5, d(c7)::5,
      d(c8)::5, d(c9)::5, d(c10)::5, d(c11)::5, d(c12)::5, d(c13)::5, d(c14)::5,
      d(c15)::5, d(c16)::5, d(c17)::5, d(c18)::5, d(c19)::5, d(c20)::5,
      d(c21)::5, d(c22)::5, d(c23)::5, d(c24)::5, d(c25)::5, d(c26)::5>>
  catch
    :error -> :error
  else
    decoded -> {:ok, decoded}
  end

  defp decode(_), do: :error

  @compile {:inline, d: 1}

  defp d(?0), do: 0
  defp d(?1), do: 1
  defp d(?2), do: 2
  defp d(?3), do: 3
  defp d(?4), do: 4
  defp d(?5), do: 5
  defp d(?6), do: 6
  defp d(?7), do: 7
  defp d(?8), do: 8
  defp d(?9), do: 9
  defp d(?A), do: 10
  defp d(?B), do: 11
  defp d(?C), do: 12
  defp d(?D), do: 13
  defp d(?E), do: 14
  defp d(?F), do: 15
  defp d(?G), do: 16
  defp d(?H), do: 17
  defp d(?J), do: 18
  defp d(?K), do: 19
  defp d(?M), do: 20
  defp d(?N), do: 21
  defp d(?P), do: 22
  defp d(?Q), do: 23
  defp d(?R), do: 24
  defp d(?S), do: 25
  defp d(?T), do: 26
  defp d(?V), do: 27
  defp d(?W), do: 28
  defp d(?X), do: 29
  defp d(?Y), do: 30
  defp d(?Z), do: 31
  defp d(_), do: throw(:error)

  defp valid?(
         <<c1::8, c2::8, c3::8, c4::8, c5::8, c6::8, c7::8, c8::8, c9::8,
           c10::8, c11::8, c12::8, c13::8, c14::8, c15::8, c16::8, c17::8,
           c18::8, c19::8, c20::8, c21::8, c22::8, c23::8, c24::8, c25::8,
           c26::8>>
       ) do
    v(c1) && v(c2) && v(c3) && v(c4) && v(c5) && v(c6) && v(c7) && v(c8) &&
      v(c9) && v(c10) &&
      v(c11) && v(c12) && v(c13) &&
      v(c14) && v(c15) && v(c16) && v(c17) && v(c18) && v(c19) && v(c20) &&
      v(c21) && v(c22) &&
      v(c23) && v(c24) && v(c25) && v(c26)
  end

  defp valid?(_), do: false

  @compile {:inline, v: 1}

  defp v(?0), do: true
  defp v(?1), do: true
  defp v(?2), do: true
  defp v(?3), do: true
  defp v(?4), do: true
  defp v(?5), do: true
  defp v(?6), do: true
  defp v(?7), do: true
  defp v(?8), do: true
  defp v(?9), do: true
  defp v(?A), do: true
  defp v(?B), do: true
  defp v(?C), do: true
  defp v(?D), do: true
  defp v(?E), do: true
  defp v(?F), do: true
  defp v(?G), do: true
  defp v(?H), do: true
  defp v(?J), do: true
  defp v(?K), do: true
  defp v(?M), do: true
  defp v(?N), do: true
  defp v(?P), do: true
  defp v(?Q), do: true
  defp v(?R), do: true
  defp v(?S), do: true
  defp v(?T), do: true
  defp v(?V), do: true
  defp v(?W), do: true
  defp v(?X), do: true
  defp v(?Y), do: true
  defp v(?Z), do: true
  defp v(_), do: false
end
